README.markdown

```diff
@@ -21,16 +21,313 @@ and then the reviews:
+
+# MySQL version
+Book.import [book], on_duplicate_key_update: [:title]
+
+# PostgreSQL version
+Book.import [book], on_duplicate_key_update: {conflict_target: [:id], columns: [:title]}
+
+# PostgreSQL shorthand version (conflict target must be primary key)
+Book.import [book], on_duplicate_key_update: [:title]
+
+book.reload.title  # => "Updated Book Title" (changed)
+book.reload.author # => "FooManChu"          (stayed the same)
+```
+
+h2. Using the value from another column
+
+```ruby
+book = Book.create! title: "Book1", author: "FooManChu"
+book.title = "Updated Book Title"
+
+# MySQL version
+Book.import [book], on_duplicate_key_update: {author: :title}
+
+# PostgreSQL version (no shorthand version)
+Book.import [book], on_duplicate_key_update: {
+  conflict_target: [:id], columns: {author: :title}
+}
+
+book.reload.title  # => "Book1"              (stayed the same)
+book.reload.author # => "Updated Book Title" (changed)
+```
+
+h2. Using Custom SQL
+
+```ruby
+book = Book.create! title: "Book1", author: "FooManChu"
+book.author = "Bob Barker"
+
+# MySQL version
+Book.import [book], on_duplicate_key_update: "author = values(author)"
+
+# PostgreSQL version
+Book.import [book], on_duplicate_key_update: {
+  conflict_target: [:id], columns: "author = excluded.author"
+}
+
+# PostgreSQL shorthand version (conflict target must be primary key)
+Book.import [book], on_duplicate_key_update: "author = excluded.author"
+
+book.reload.title  # => "Book1"      (stayed the same)
+book.reload.author # => "Bob Barker" (changed)
+```
+
+h2. PostgreSQL Using constraints
+
+```ruby
+book = Book.create! title: "Book1", author: "FooManChu", edition: 3, published_at: nil
+book.published_at = Time.now
+
+# in migration
+execute <<-SQL
+      ALTER TABLE books
+        ADD CONSTRAINT for_upsert UNIQUE (title, author, edition);
+    SQL
+
+# PostgreSQL version
+Book.import [book], on_duplicate_key_update: {constraint_name: :for_upsert, columns: [:published_at]}
+
+
+book.reload.title  # => "Book1"          (stayed the same)
+book.reload.author # => "FooManChu"      (stayed the same)
+book.reload.edition # => 3               (stayed the same)
+book.reload.published_at # => 2017-10-09 (changed)
+```
+
+#### Uniqueness Validation
+
+By default, `activerecord-import` will not validate for uniquness when importing records. Starting with `v0.27.0`, there is a  parameter called `validate_uniqueness` that can be passed in to trigger this behavior. This option is provided with caution as there are many potential pitfalls. Please use with caution.
+
+```ruby
+Book.import books, validate_uniqueness: true
+```
+
+### Array of Hashes
+
+Due to the counter-intuitive behavior that can occur when dealing with hashes instead of ActiveRecord objects, `activerecord-import` will raise an exception when passed an array of hashes. If you have an array of hash attributes, you should instead use them to instantiate an array of ActiveRecord objects and then pass that into `import`.
```

Comment:
@jkowens @ahmad-sanad I want to confirm that I understood the issue here correctly.

Reply:
So an array of hashes does work, but only if the columns are consistent in every hash in the array.
e.g.
```
MyModel.import [{foo: 1}, {foo: 2, bar: 3}] # raises an exception

MyModel.import [{foo: 1, bar: 2}, {foo: 3, bar: 4}] # works fine
```

The potential workarounds discussed were:

1. Convert it to an array of ActiveRecord objects
```
# Instead of this
MyModel.import [{foo: 1}, {foo: 2, bar: 3}]

# Do this
MyModel.import [MyModel.new(foo: 1), MyModel.new(foo: 2, bar: 3)]

# but hopefully with cleaner code lol
```

2. Divide the array into multiple ones with consistent columns and import each one separately
```
# Instead of this
MyModel.import [{foo: 1}, {foo: 2, bar: 3}] 

# Do this
MyModel.import [{foo: 1}] 
MyModel.import [{foo: 2, bar: 3}] 
```

Reply:
Also, thanks for adding this!!!

Reply:
NP! Thanks for the clarification. I will fix this up.

Reply:
So in regards to no. 2 the thought was to make use of Enumerable's `#group_by`.

For example:

```
arr = [{foo: 1}, {foo: 2, bar: 3}] 
arr.group_by(&:keys).each_value do |v|
 MyModel.import v
end
```
I think this would be the recommended workaround, but I don't have any benchmark data to support that.

Reply:
Updated hash examples again with `group_by` @jkowens